### PR TITLE
Add analytics support when generating URL

### PR DESCRIPTION
**Pull request by adimiz1**
Add a support for SDK analytics
If query parameters already exist do not add analytics suffix.

### CODE REVIEW

File: cloudinary-core/src/main/java/com/cloudinary/Url.java [hunk 1]
```diff
@@ -417,6 +418,9 @@ public String generate(String source) {
             } catch (MalformedURLException ignored) {
             }
         }
+        if (AnalyticsUtils.token != null) {
+            url = (new StringBuilder()).append(url).append(AnalyticsUtils.analyticsPrefix).append(AnalyticsUtils.token).toString();
```
Comment (patrick-tolosa) [inline cloudinary-core/src/main/java/com/cloudinary/Url.java:422]: The URL can already include a query param, if the URL already includes a query param, we should not append the analytics token.

**Review by patrick-tolosa — CHANGES_REQUESTED** [review]
Please add a description in the PR of the changes here.

What was added?, how does it affect the user?  in what context can it be used?

File: cloudinary-core/src/main/java/com/cloudinary/Url.java (hunk 1 above)
Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:422]: Fixing

**Comment by adimiz1** [comment]
Add a support for SDK analytics
If query parameters already exist do not add analytics suffix.

**Review by patrick-tolosa — APPROVED** [review]

File: cloudinary-core/src/main/java/com/cloudinary/Url.java [hunk 2]
```diff
@@ -417,6 +418,11 @@ public String generate(String source) {
             } catch (MalformedURLException ignored) {
             }
         }
+        if (AnalyticsUtils.token != null) {
+            if(!AnalyticsUtils.checkIfQueryParamExist(url)) {
+                url = (new StringBuilder()).append(url).append(AnalyticsUtils.analyticsPrefix).append(AnalyticsUtils.token).toString();
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/Url.java:423]: Why don't we use `URL` class(like with  `authToken` above) and perform less safe string manipulations?

**Review by const-cloudinary — CHANGES_REQUESTED** [review]
@adimiz1, please see review comments,

In addition, PR title is a bit ambiguous, in Java SDK there is no such functionality called `url-gen`

File: cloudinary-core/src/main/java/com/cloudinary/Cloudinary.java [hunk 3]
```diff
@@ -153,6 +154,15 @@ public void setUserAgent(String prefix, String version){
         userAgent = prefix+"/"+ version + " ("+USER_AGENT_PREFIX+ " "+VERSION+") " + USER_AGENT_JAVA_VERSION;
     }
 
+    /**
+     * Set the prefix and version for the user agent that will be sent with every API call
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/Cloudinary.java:158]: Please fix the doc string

File: cloudinary-core/src/main/java/com/cloudinary/Url.java [hunk 4]
```diff
@@ -408,6 +409,7 @@ public String generate(String source) {
         String join = StringUtils.join(new String[]{prefix, finalResourceType, signature, transformationStr, version, source}, "/");
         String url = StringUtils.mergeSlashesInUrl(join);
 
+
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/Url.java:412]: redundant line?

File: cloudinary-core/src/main/java/com/cloudinary/Url.java [hunk 5]
```diff
@@ -417,6 +419,16 @@ public String generate(String source) {
             } catch (MalformedURLException ignored) {
             }
         }
+        if (cloudinary.analytics != null && cloudinary.config.analytics) {
+            try {
+                URL tempUrl = new URL(url);
+                if (tempUrl.getQuery() == null) {
+                    String path = tempUrl.getPath();
+                    url = url + "?" + cloudinary.analytics.toQueryParam();
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/Url.java:427]: what if we have both `authToken` and `analytics`? we need to make sure we build the right query string( using `&`)

File: cloudinary-core/src/main/java/com/cloudinary/utils/Analytics.java [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        if (SDKSemver.split("\\.").length < 2) {
+            throw new Exception("invalid semVer, must have at least two segments");
+        }
+        String[] versionArray = SDKSemver.split("\\.");
+        for (int i = 0 ; i < versionArray.length; i ++) {
+            versionArray[i] = StringUtils.padStart(versionArray[i], 2, '0');
+        }
+        return StringUtils.join(StringUtils.reverseStringArray(versionArray), ".");
+    }
+}
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/utils/Analytics.java:113]: no newline at the end of file

File: cloudinary-core/src/main/java/com/cloudinary/utils/Base64Map.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+import java.util.HashMap;
+import java.util.Map;
+
+public class Base64Map {
+    public static Map<String, String> values;
+
+    static {
+        values = new HashMap<>();
+        values.put("000000", "A");
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/utils/Base64Map.java:11]: no better way to populate a hashmap in Java?

File: cloudinary-core/src/main/java/com/cloudinary/utils/StringUtils.java [hunk 8]
```diff
@@ -397,4 +396,55 @@ public static String mergeSlashesInUrl(String url) {
...
+    }
+
+    /**
+     * Break string into groups of n size strings
+     *
+     * @param text The string to process
+     * @param n Size of group
+     * @return List with all strings with group size n.
+     */
+    public static List<String> usingPattern(String text, int n) {
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/utils/StringUtils.java:439]: The name of the function is kind of ambiguous...

File: cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java [hunk 9]
```diff
@@ -92,6 +93,16 @@ public void testAuthenticatedUrl() {
 
     }
 
+    @Test
+    public void testUrlAnalyticsWithQueryParams() {
+        cloudinary.config.analytics = true;
+        cloudinary.setAnalytics(new Analytics("F", "2.0.0", System.getProperty("java.version")));
+        cloudinary.config.privateCdn = true;
+        String url = cloudinary.url().signed(true).type("authenticated").generate("test");
+        assertEquals(url,"http://test123-res.cloudinary.com/image/authenticated/test?__cld_token__=st=11111111~exp=11111411~hmac=735a49389a72ac0b90d1a84ac5d43facd1a9047f153b39e914747ef6ed195e53");
```
Comment (const-cloudinary) [inline cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java:102]: where is the analytics part in the URL?

File: cloudinary-core/src/main/java/com/cloudinary/utils/Analytics.java [hunk 10]
```diff
@@ -0,0 +1,113 @@
+package com.cloudinary.utils;
+
+import java.util.Arrays;
+import java.util.List;
+
+public class Analytics {
+    private String sdkTokenQueryKey = "_a="; //sdkTokenQueryKey
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/utils/Analytics.java:7]: The query key is actually `_a`, `=` is the delimiter between key and value.

File: cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+        Assert.assertEquals(result, "_a=AFAACMh0");
+    }
+
+    @Test
+    public void testUrlWithAnalytics() {
+        cloudinary.config.analytics = true;
+        cloudinary.setAnalytics(new Analytics("F", "2.0.0", "1.8.0"));
+        String url = cloudinary.url().generate("test");
+        Assert.assertEquals(url, "http://res.cloudinary.com/test123/image/upload/test?_a=AFAACMh0");
+        cloudinary.config.analytics = false;
```
Comment (const-cloudinary) [inline cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java:37]: this (and all the other occurrences) should be set in the `tearDown` method. If the assertion fails, this code is never executed. This state might also affect other tests.

File: java_shared.gradle [hunk 12]
```diff
@@ -2,6 +2,7 @@ sourceCompatibility = 1.7
 targetCompatibility = 1.7
 
 javadoc {
+    failOnError false
```
Comment (const-cloudinary) [inline java_shared.gradle:5]: is this related to the PR?

**Review by const-cloudinary — CHANGES_REQUESTED** [review]
@adimiz1 , a very good PR! Please fix a few minor issues.

File: cloudinary-core/src/main/java/com/cloudinary/Url.java (hunk 4 above)
Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:412]: Not as far as I know (this was not added by me)
join = http://test123-res.cloudinary.com/image/authenticated////test
url = http://test123-res.cloudinary.com/image/authenticated/test
So I think both functions are required. 
Not sure yet why join has all those extra '/'

File: cloudinary-core/src/main/java/com/cloudinary/Url.java (hunk 5 above)
Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:427]: As part of this ticket: https://cloudinary.atlassian.net/browse/SNI-5996
I needed to prevent the analytics for every URL that has query params, including auth token.

File: cloudinary-core/src/main/java/com/cloudinary/utils/Analytics.java (hunk 6 above)
Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/utils/Analytics.java:113]: Not sure I understand, each file suppose to have a new line at the end of it?

File: cloudinary-core/src/main/java/com/cloudinary/utils/Base64Map.java (hunk 7 above)
Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/utils/Base64Map.java:11]: No :(, Java 6, this was a nightmare
With Java 8 this could have been done much easier.

File: cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java (hunk 9 above)
Comment (adimiz1) [reply cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java:102]: Since every url with query params will not have analytics I've removed it from this test as well

File: java_shared.gradle (hunk 12 above)
Comment (adimiz1) [reply java_shared.gradle:5]: Yes
I'm still trying to figure why the javadoc fails the build.

File: cloudinary-core/src/main/java/com/cloudinary/Url.java (hunk 4 above)
Comment (patrick-tolosa) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:412]: I think he was talking about line 412, that's empty :)

File: cloudinary-core/src/main/java/com/cloudinary/Url.java [hunk 13]
```diff
@@ -417,6 +419,16 @@ public String generate(String source) {
             } catch (MalformedURLException ignored) {
             }
         }
+        if (cloudinary.analytics != null && cloudinary.config.analytics) {
```
Comment (patrick-tolosa) [inline cloudinary-core/src/main/java/com/cloudinary/Url.java:422]: I'm not sure this line is clear enough.

What's the difference between `cloudinary.analytics` and `cloudinary.config.analytics`?
Why do we both conditions need to be true?

File: cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java [hunk 14]
```diff
@@ -92,6 +93,15 @@ public void testAuthenticatedUrl() {
 
     }
 
+    @Test
+    public void testUrlAnalyticsWithQueryParams() {
+        cloudinary.config.privateCdn = true;
+        String url = cloudinary.url().signed(true).type("authenticated").generate("test");
+        assertEquals(url,"http://test123-res.cloudinary.com/image/authenticated/test?__cld_token__=st=11111111~exp=11111411~hmac=735a49389a72ac0b90d1a84ac5d43facd1a9047f153b39e914747ef6ed195e53");
```
[the remark this answers is not in the document]
Comment (patrick-tolosa) [reply cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java:102]: As discussed in other conversations, we'll be strict here and assume that analytics can only work when there are no other query params.

**Review by patrick-tolosa — APPROVED** [review]
I'm leaving @const-cloudinary to resolve with you all the open issues, PR looks good, it needs some finishing touches.

File: cloudinary-core/src/main/java/com/cloudinary/Url.java (hunk 13 above)
Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:422]: This was part of the design by @const-cloudinary and I
Think of the situation where we set the analytics object by default from the Android by the developer explicitly does not want to send analytics
@const-cloudinary suggest we do this as part of the config we init, so if the develop is running it through some kind of YML, XML or any other config file he can just set analytics = false

Comment (const-cloudinary) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:422]: `cloudinary.config.analytics` is the configuration(boolean) for turning analytics on and off. `cloudinary.analytics` is the Analytics class itself that generates the token. It can be configured from the outside (android) with the relevant SDK key and other things.

Comment (const-cloudinary) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:422]: And I agree with @patrick-tolosa, that class should be always initialized in the constructor (whether in Java SDK or in Android)

File: cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java (hunk 9 above)
Comment (const-cloudinary) [reply cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java:102]: maybe let's update the unit test name to something like `testUrlNoAnalyticsWithQueryParams `

File: cloudinary-core/src/main/java/com/cloudinary/Cloudinary.java [hunk 15]
```diff
@@ -153,6 +154,15 @@ public void setUserAgent(String prefix, String version){
         userAgent = prefix+"/"+ version + " ("+USER_AGENT_PREFIX+ " "+VERSION+") " + USER_AGENT_JAVA_VERSION;
     }
 
+    /**
+     * Set the analytics object that will be sent with every URL generation call.
+     * a userAgent is built from `prefix/version (additional data)`
```
Comment (const-cloudinary) [inline cloudinary-core/src/main/java/com/cloudinary/Cloudinary.java:159]: userAgent? how is it related to the analytics?

File: cloudinary-core/src/main/java/com/cloudinary/Url.java (hunk 5 above)
Comment (const-cloudinary) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:427]: Where can I see that in code?
I see this line:
`if (signUrl && authToken != null && !authToken.equals(AuthToken.NULL_AUTH_TOKEN)) {`
and this line:
`if (cloudinary.analytics != null && cloudinary.config.analytics) {`

They both can be evaluated to `true`.

File: cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+public class AnalyticsTest {
+
+    private Cloudinary cloudinary;
+
+    @Rule
+    public TestName currentTest = new TestName();
+
+    @Before
+    public void setUp() {
+        System.out.println("Running " + this.getClass().getName() + "." + currentTest.getMethodName());
```
Comment (const-cloudinary) [inline cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java:17]: Do we really need to print things here?

File: cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+        Assert.assertEquals(result, "_a=E");
+    }
+
+    @After
+    public void tearDown() {
+        cloudinary.config.analytics = false;
+        cloudinary.analytics = null;
+    }
+
+}
```
Comment (const-cloudinary) [inline cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java:86]: No newline at end of file

**Review by const-cloudinary — CHANGES_REQUESTED** [review]

File: cloudinary-core/src/main/java/com/cloudinary/Url.java (hunk 13 above)
Comment (patrick-tolosa) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:422]: The condition seems a bit off for anyone not in the details.
a flag of "on and off" is already called `analytics` in other SDKs (A bit unfortunate), but the perhaps we can find a better name for the class/instance?

Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:422]: So according to what I read, we should init the analytics in the constructor (with Java params)
And we want to change hte name of the class `analytics`? Let me know what name you want to put instead

File: cloudinary-core/src/main/java/com/cloudinary/Cloudinary.java (hunk 15 above)
Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/Cloudinary.java:159]: It's not , it's the function above the `setAnalytics` function

Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/Cloudinary.java:159]: Here i just changed the documentation for the `setAnalytics` function.

File: cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java (hunk 16 above)
Comment (adimiz1) [reply cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java:17]: No, removing it.

File: java_shared.gradle (hunk 12 above)
Comment (adimiz1) [reply java_shared.gradle:5]: removed.

File: cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java (hunk 17 above)
Comment (adimiz1) [reply cloudinary-core/src/test/java/com/cloudinary/analytics/AnalyticsTest.java:86]: fixed

File: cloudinary-core/src/main/java/com/cloudinary/Url.java (hunk 4 above)
Comment (adimiz1) [reply cloudinary-core/src/main/java/com/cloudinary/Url.java:412]: Fixing

File: cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java (hunk 9 above)
Comment (adimiz1) [reply cloudinary-core/src/test/java/com/cloudinary/AuthTokenTest.java:102]: Done

**Review by const-cloudinary — APPROVED** [review]
